### PR TITLE
add ping pong for tri protocol

### LINKED ISSUE #12945: 服务端突然断网、断电，客户端无法进行故障转移
dubbo 版本 3.2.4

实验过程：服务端关闭网络，节点在15min内不会从nacos平台消失。导致consumer仍然能够拿到invoder，此时客户端如果能判断服务端invoker不能连接，就会从全局的invoker列表中剔除该invoker，执行故障转移。但是dubbo consumer 在服务端突然断网后，判断连接仍然存在，导致无法剔除invoker，无法进行故障转移。

问题：针对服务端突然断电、断网，是否是应为tcp链接无法正常断开，导致consumer判断连接失效？

image

### PR DESCRIPTION
**Pull request by icodening**
## What is the purpose of the change
fix #12945
add ping pong for tri protocol

## Brief changelog
add fixed delay write ping frame handler

## Checklist
- [ ] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleWritePingHandler.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    private ScheduledFuture<?> scheduledFuture;
+
+    public TripleWritePingHandler(long writePingIntervalInMs) {
+        this.writePingIntervalInMs = writePingIntervalInMs;
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        super.channelActive(ctx);
+        scheduledFuture = ctx.executor().scheduleWithFixedDelay(new WritePingTask(ctx.channel()), writePingIntervalInMs, writePingIntervalInMs, TimeUnit.MILLISECONDS);
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleWritePingHandler.java:42]: Need a better way to write heartbeat.

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-remoting/dubbo-remoting-netty4/src/main/java/org/apache/dubbo/remoting/transport/netty4/NettyConnectionClient.java [hunk 2]
```diff
@@ -122,12 +125,15 @@ protected void initChannel(SocketChannel ch) {
                 }
 
 //                pipeline.addLast("logging", new LoggingHandler(LogLevel.INFO)); //for debug
-                // TODO support IDLE
-//                int heartbeatInterval = UrlUtils.getHeartbeat(getUrl());
+
+                int idleTimeout = UrlUtils.getIdleTimeout(getUrl());
+                pipeline.addLast("client-idle-handler", new IdleStateHandler(idleTimeout, 0, 0, MILLISECONDS));
```
Comment (AlbumenJ) [inline dubbo-remoting/dubbo-remoting-netty4/src/main/java/org/apache/dubbo/remoting/transport/netty4/NettyConnectionClient.java:130]: Need a idle handler and a close handler.

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleWritePingHandler.java (hunk 1 above)
Comment (icodening) [reply dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleWritePingHandler.java:42]: ok , i will redesign it

**Review by EarthChen — APPROVED** [review]
LGTM

**Comment by AlbumenJ** [comment]
If we use port unification for both Dubbo and Triple protocol, will it work expectedly?

**Comment by icodening** [comment]
> If we use port unification for both Dubbo and Triple protocol, will it work expectedly?

Only Triple protocol will handle `IDLE` state

**Comment by AlbumenJ** [comment]
> > If we use port unification for both Dubbo and Triple protocol, will it work expectedly?
> 
> Only Triple protocol will handle `IDLE` state

But Dubbo protocol alose use `IDLE` state too

**Comment by AlbumenJ** [comment]
@guohao pPTAL

**Review by guohao — APPROVED** [review]

**Comment by guohao** [comment]
> > If we use port unification for both Dubbo and Triple protocol, will it work expectedly?
> 
> Only Triple protocol will handle `IDLE` state

Dubbo protocol's `IDEL` event can be supported in a seperated PR

**Pull request merged**